pkg/openstack/common.go

```diff
@@ -27,3 +40,232 @@ func EnsureDeleted(ctx context.Context, helper *helper.Helper, obj client.Object
+}
+
+// GetEndpointURL - returns the URL of the endpoint (proto://hostname/path)
+func (sd *ServiceDetails) GetEndpointURL() string {
+	return sd.endpointURL
+}
+
+// public endpoint
+// 1) Default (service override nil)
+// - create route + pass in service selector
+// 2) service override provided == LoadBalancer service type
+// - do nothing let the service operator handle it
+// 3) service override provided == ClusterIP service type
+// - create route , merge override
+//
+// internal endpoint
+// 1) Default (service override nil)
+// - service operator will create default service
+// 2) service override provided
+// - do nothing let the service operator handle it
+
+// CreateEndpointServiceOverride -
+func (sd *ServiceDetails) CreateEndpointServiceOverride() (*service.Service, error) {
+
+	// get service overrides for the endpoint
+	svcOverride := ptr.To(sd.ServiceOverrideSpec[string(sd.Endpoint)])
+
+	// set the endpoint name to service name - endpoint type
+	sd.endpointName = sd.ServiceName + "-" + string(sd.Endpoint)
+
+	// if there is no override for the internal endpoint return
+	if svcOverride == nil && sd.Endpoint == service.EndpointInternal {
+		return nil, nil
+	}
+
+	// initialize the service override if not specified via the CR
+	if svcOverride == nil {
+		svcOverride = &service.OverrideSpec{}
+	}
+
+	// Create generic service definitions for either MetalLB, or generic service
+	svcDef := &k8s_corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      sd.endpointName,
+			Namespace: sd.Namespace,
+		},
+		Spec: k8s_corev1.ServiceSpec{
+			Type: k8s_corev1.ServiceTypeClusterIP,
+		},
+	}
+
+	// Create the service
+	svc, err := service.NewService(
+		svcDef,
+		time.Duration(5)*time.Second, //not really required as we don't create the service here
+		svcOverride,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// add annotation to register service name in dnsmasq if it is a LoadBalancer service
+	if svc.GetServiceType() == k8s_corev1.ServiceTypeLoadBalancer {
+		svc.AddAnnotation(map[string]string{
+			service.AnnotationHostnameKey: svc.GetServiceHostname(),
+		})
+	}
+
+	return svc, nil
+}
+
+// CreateRoute -
+func (sd *ServiceDetails) CreateRoute(
+	ctx context.Context,
+	helper *helper.Helper,
+	svc service.Service,
+) (ctrl.Result, error) {
+	// TODO TLS
+	route, err := route.NewRoute(
+		route.GenericRoute(&route.GenericRouteDetails{
+			Name:      sd.ServiceName,
+			Namespace: sd.Namespace,
+			Labels: map[string]string{
+				common.AppSelector: sd.ServiceName,
+			},
+			ServiceName:    sd.endpointName,
+			TargetPortName: sd.endpointName,
+		}),
+		time.Duration(5)*time.Second,
+		sd.RouteOverrideSpec,
+	)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	ctrlResult, err := route.CreateOrPatch(ctx, helper)
+	if err != nil {
+		return ctrlResult, err
+	} else if (ctrlResult != ctrl.Result{}) {
+		return ctrlResult, nil
+	}
+
+	sd.hostname = ptr.To(route.GetHostname())
+	sd.endpointURL = "http://" + *sd.hostname
```

Comment:
> If I understand the resulting CRD model and the code correctly then KeystoneAPISpec.override["public"].endpointURL will be unconditionally populated with the information generated here.
> 
> I think this means that OpenStackControlPlane.spec.keystone.template.override[].endpointURL is a field in the CRD that should never be set by the human creating the OpenStackControlPlane CR. I think the best would be not to have such a spec field that is not intended to be ever set.
> 
> In the Nova CRD we intentionally not include the service CRD Spec structs directly to NovaSpec but have a parallel template struct. This way we can keep the fields with generated value out of the struct used by the user (the template) but keep it in the struct used by the service reconciler (spec).
> 
> If we want to apply that pattern here to hide the unsettable endpointURL from the OpenStackControlPlane then the OpenStackControlPlaneSpec should stop including the KeystoneAPISpec. Instead a KeystoneAPITemplate struct needs to be created that has all the KeystoneAPISpec fields except endpointURL. The value of that field is generated by the openstack-operator so it can provided that extra value when it creates the KeystoneAPISpec from the KeystoneAPITemplate.
> 
> (of course KeystoneAPISpec is just an example here, this comment is valid for all the endpointURL fields being introduced now)

The idea is that a human/user can set it if the intend is to use it with an ingress other then an OCP route (not implemented in this patch). when endpointURL gets set by a user no route gets created, and just the endpointURL passed into the service operators, to be used for the service registration in keystone. Then the user has to create the ingress manually.

Reply:
Do we really want to support both use cases (route creation by the openstack operator, and endpoint url passthrough)? Does it mean the user can first configure OpenStackControlPlane with automatic route creation then later specify the endpoint url manually and therefore instruct the openstack-operator to delete the route, and instruct the service operator to update the endpoint url in keystone according the new new user specified value? The reverse direction too to switch from manually created route to openstack-operator created route?